src/librbd/WatchNotifyTypes.h

```diff
@@ -67,299 +68,433 @@ enum NotifyOp {
-  void dump(Formatter *f) const;
+  NotifyOp get_notify_op() const override {
+    return NOTIFY_OP_SPARSIFY;
+  }
+  bool check_for_refresh() const override {
+    return true;
+  }
+
+  void encode(bufferlist &bl) const override;
+  void decode(__u8 version, bufferlist::const_iterator &iter) override;
+  void dump(Formatter *f) const override;
 };
 
-struct UnknownPayload {
-  static const NotifyOp NOTIFY_OP = static_cast<NotifyOp>(-1);
-  static const bool CHECK_FOR_REFRESH = false;
+struct QuiescePayload : public AsyncRequestPayloadBase {
+  QuiescePayload() {}
+  QuiescePayload(const AsyncRequestId &id) : AsyncRequestPayloadBase(id) {}
 
-  void encode(bufferlist &bl) const;
-  void decode(__u8 version, bufferlist::const_iterator &iter);
-  void dump(Formatter *f) const;
+  virtual ~QuiescePayload() {}
+
+  NotifyOp get_notify_op() const override {
+    return NOTIFY_OP_QUIESCE;
+  }
+  bool check_for_refresh() const override {
+    return false;
+  }
 };
 
-typedef boost::variant<AcquiredLockPayload,
-                       ReleasedLockPayload,
-                       RequestLockPayload,
-                       HeaderUpdatePayload,
-                       AsyncProgressPayload,
-                       AsyncCompletePayload,
-                       FlattenPayload,
-                       ResizePayload,
-                       SnapCreatePayload,
-                       SnapRemovePayload,
-                       SnapRenamePayload,
-                       SnapProtectPayload,
-                       SnapUnprotectPayload,
-                       RebuildObjectMapPayload,
-                       RenamePayload,
-                       UpdateFeaturesPayload,
-                       MigratePayload,
-                       SparsifyPayload,
-                       UnknownPayload> Payload;
+struct UnquiescePayload : public AsyncRequestPayloadBase {
+  UnquiescePayload() {}
+  UnquiescePayload(const AsyncRequestId &id) : AsyncRequestPayloadBase(id) {}
+
+  virtual ~UnquiescePayload() {}
+
+  NotifyOp get_notify_op() const override {
+    return NOTIFY_OP_UNQUIESCE;
+  }
+  bool check_for_refresh() const override {
+    return false;
+  }
+};
+
+struct UnknownPayload : public Payload {
+  virtual ~UnknownPayload() {}
+
+  NotifyOp get_notify_op() const override {
+    return static_cast<NotifyOp>(-1);
+  }
+  bool check_for_refresh() const override {
+    return false;
+  }
+
+  void encode(bufferlist &bl) const override;
+  void decode(__u8 version, bufferlist::const_iterator &iter) override;
+  void dump(Formatter *f) const override;
+};
 
 struct NotifyMessage {
-  NotifyMessage() : payload(UnknownPayload()) {}
-  NotifyMessage(const Payload &payload_) : payload(payload_) {}
+  NotifyMessage() : payload(new UnknownPayload()) {}
+
+  NotifyMessage(Payload *payload) : payload(payload) {}
+
+  NotifyMessage(const NotifyMessage& m) {
+    *this = m;
+  }
+
+  NotifyMessage &operator=(const NotifyMessage& m);
```

Comment:
We don't need the assignment operator for librbd, but for dencoder. Compiling ceph-dencoder fails with:
```
/home/mgolub/ceph/ceph/src/tools/ceph-dencoder/denc_registry.h:113:8: error: use of deleted function ‘librbd::watch_notify::NotifyMessage& librbd::watch_notify::NotifyMessage::operator=(const librbd::watch_notify::NotifyMessage&)’
  113 |     *n = *this->m_object;
      |     ~~~^~~~~~~~~~~~~~~~~
...
/home/mgolub/ceph/ceph/src/librbd/WatchNotifyTypes.h:481:8: note: ‘librbd::watch_notify::NotifyMessage& librbd::watch_notify::NotifyMessage::operator=(const librbd::watch_notify::NotifyMessage&)’ is implicitly deleted because the default definition would be ill-formed:
  481 | struct NotifyMessage {
      |        ^~~~~~~~~~~~~
/home/mgolub/ceph/ceph/src/librbd/WatchNotifyTypes.h:481:8: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = librbd::watch_notify::Payload; _Dp = std::default_delete<librbd::watch_notify::Payload>]’
```
and also:
```
/home/mgolub/ceph/ceph/src/tools/ceph-dencoder/denc_registry.h:118:12: error: use of deleted function ‘librbd::watch_notify::NotifyMessage::NotifyMessage(const librbd::watch_notify::NotifyMessage&)’
  118 |     T *n = new T(*this->m_object);
      |            ^~~~~~~~~~~~~~~~~~~~~~
```
So, the problem is using the `unique_ptr`, due to which I need to explicitly implement the assignment operator and copy constructor.

Reply:
Could you switch the `TYPE(librbd::watch_notify::NotifyMessage)` to `TYPE_NOCOPY(librbd::watch_notify::NotifyMessage)`?

Reply:
indeed. Thanks